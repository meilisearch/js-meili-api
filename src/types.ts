// Type definitions for meilisearch
// Project: https://github.com/meilisearch/meilisearch-js
// Definitions by: qdequele <quentin@meilisearch.com> <https://github.com/meilisearch>
// Definitions: https://github.com/meilisearch/meilisearch-js
// TypeScript Version: ^3.8.3

import {
  AxiosError,
  AxiosInstance,
  AxiosRequestConfig,
  AxiosResponse,
  CancelTokenSource,
} from 'axios'

import { Index } from './index'
import MeiliAxiosWrapper from './meili-axios-wrapper'
import MeiliSearch from './meilisearch'
import MeiliSearchApiError from './errors/meilisearch-api-error'
import MeiliSearchTimeOutError from './errors/meilisearch-timeout-error'
export { Index }
export { MeiliSearchApiError }
export { MeiliSearchTimeOutError }

export interface Config {
  host: string
  apiKey?: string
}

///
/// Request specific interfaces
///

export interface IndexRequest {
  uid: string
  primaryKey?: string
}

export interface IndexOptions {
  primaryKey?: string
}

export interface IndexResponse {
  uid: string
  name?: string
  primaryKey?: string
  createdAt: Date
  updatedAt: Date
}

export interface AddDocumentParams {
  primaryKey?: string
}

export type FacetFilter = (string | string[])[]

export interface SearchParams<T> {
  offset?: number
  limit?: number
  attributesToRetrieve?: Extract<keyof T, string>[] | Extract<keyof T, string>
  attributesToCrop?:
    | (Extract<keyof T, string> | '*')[]
    | (Extract<keyof T, string> | '*')
  cropLength?: number
  attributesToHighlight?:
    | (Extract<keyof T, string> | '*')[]
    | (Extract<keyof T, string> | '*')
  filters?: string
  facetFilters?: string | FacetFilter | FacetFilter[]
  facetsDistribution?: string[]
  matches?: boolean
}

export interface SearchRequest {
  q: string
  offset?: number
  limit?: number
  attributesToRetrieve?: string
  attributesToCrop?: string
  cropLength?: number
  attributesToHighlight?: string
  facetFilters?: string
  facetsDistribution?: string
  filters?: string
  matches?: boolean
}

export type Hit<T> = T & { _formatted?: T }

export interface SearchResponse<T, P extends SearchParams<T>> {
  hits: P['attributesToRetrieve'] extends keyof T
    ? Array<
        Hit<
          Pick<T, Exclude<keyof T, Exclude<keyof T, P['attributesToRetrieve']>>>
        >
      >
    : P['attributesToRetrieve'] extends Array<infer K>
    ? Array<Hit<Pick<T, Exclude<keyof T, Exclude<keyof T, K>>>>>
    : Array<Hit<T>>
  offset: number
  limit: number
  processingTimeMs: number
  facetsDistribution?: object
  exhaustiveFacetsCount?: boolean
  query: string
}

export interface FieldFrequency {
  [field: string]: number
}

/*
 ** Documents
 */
export interface GetDocumentsParams<T> {
  offset?: number
  limit?: number
  attributesToRetrieve?: Extract<keyof T, string>[] | Extract<keyof T, string>
}

export type GetDocumentsResponse<
  T,
  P extends GetDocumentsParams<T>
> = P['attributesToRetrieve'] extends keyof T
  ? Array<
      Document<
        Pick<T, Exclude<keyof T, Exclude<keyof T, P['attributesToRetrieve']>>>
      >
    >
  : P['attributesToRetrieve'] extends Array<infer K>
  ? Array<Document<Pick<T, Exclude<keyof T, Exclude<keyof T, K>>>>>
  : Array<Document<T>>

export type DocumentLike = { [Key in string]?: DocumentField }
export interface DocumentArray extends Array<DocumentField> {}
export type DocumentField =
  | string
  | number
  | boolean
  | null
  | DocumentLike
  | DocumentArray

export type Document<T> = DocumentLike &
  {
    [key in keyof T]: T[key]
  }

/*
 ** Settings
 */

export interface Settings {
  attributesForFaceting?: string[]
  distinctAttribute?: string
  searchableAttributes?: string[]
  displayedAttributes?: string[]
  rankingRules?: string[]
  stopWords?: string[]
  synonyms?: {
    [field: string]: string[]
  }
  acceptNewFields?: boolean
}

/*
 ** UPDATE
 */

export interface EnqueuedUpdate {
  updateId: number
}

export interface Update {
  status: string
  updateId: number
  type: {
    name: string
    number: number
  }
  duration: number
  enqueuedAt: string
  processedAt: string
}

/*
 *** STATS
 */

export interface IndexStats {
  numberOfDocuments: number
  isIndexing: boolean
  fieldsFrequency: FieldFrequency
}

export interface Stats {
  databaseSize: number
  lastUpdate: string
  indexes: {
    [index: string]: IndexStats
  }
}

/*
 ** Keys
 */

export interface Keys {
  private: string | null
  public: string | null
}

/*
 ** version
 */
export interface Version {
  commitSha: string
  buildDate: string
  pkgVersion: string
}

/*
 ** SYS-INFO
 */
export interface SysInfo {
  memoryUsage: number
  processorUsage: number[]
  global: {
    totalMemory: number
    usedMemory: number
    totalSwap: number
    usedSwap: number
    inputData: number
    outputData: number
  }
  process: {
    memory: number
    cpu: number
  }
}

export interface SysInfoPretty {
  memoryUsage: string
  processorUsage: string[]
  global: {
    totalMemory: string
    usedMemory: string
    totalSwap: string
    usedSwap: string
    inputData: string
    outputData: string
  }
  process: {
    memory: string
    cpu: string
  }
}

/*
 ** MeiliSearch Class Interfaces
 */

export interface MeiliSearchInterface extends MeiliAxiosWrapper {
  config: Config
  getIndex: <T>(indexUid: string) => Index<T>
  getOrCreateIndex: <T>(
    uid: string,
    options?: IndexOptions
  ) => Promise<Index<T>>
  listIndexes: () => Promise<IndexResponse[]>
  createIndex: <T>(uid: string, options?: IndexOptions) => Promise<Index<T>>
  getKeys: () => Promise<Keys>
  isHealthy: () => Promise<boolean>
  setHealthy: () => Promise<void>
  setUnhealthy: () => Promise<void>
  changeHealthTo: (health: boolean) => Promise<void>
  stats: () => Promise<Stats>
  version: () => Promise<Version>
  sysInfo: () => Promise<SysInfo>
  prettySysInfo: () => Promise<SysInfoPretty>
}

export interface IndexInterface<T = any> extends MeiliAxiosWrapperInterface {
  uid: string
  getUpdateStatus: (updateId: number) => Promise<Update>
  getAllUpdateStatus: () => Promise<Update[]>
  search: <P extends SearchParams<T>>(
    query: string,
    options?: P
  ) => Promise<SearchResponse<T, P>>
  show: () => Promise<IndexResponse>
  updateIndex: (indexData: IndexOptions) => Promise<IndexResponse>
  deleteIndex: () => Promise<string>
  getStats: () => Promise<IndexStats>
  getDocuments: <P extends GetDocumentsParams<T>>(
    options?: P
  ) => Promise<GetDocumentsResponse<T, P>>
  getDocument: (documentId: string | number) => Promise<Document<T>>
  addDocuments: (
    documents: Array<Document<T>>,
    options?: AddDocumentParams
  ) => Promise<EnqueuedUpdate>
  updateDocuments: (
    documents: Array<Document<T>>,
    options?: AddDocumentParams
  ) => Promise<EnqueuedUpdate>
  deleteDocument: (documentId: string | number) => Promise<EnqueuedUpdate>
  deleteDocuments: (
    documentsIds: string[] | number[]
  ) => Promise<EnqueuedUpdate>
  deleteAllDocuments: () => Promise<EnqueuedUpdate>
  getSettings: () => Promise<Settings>
  updateSettings: (settings: Settings) => Promise<EnqueuedUpdate>
  resetSettings: () => Promise<EnqueuedUpdate>
  getSynonyms: () => Promise<object>
  updateSynonyms: (synonyms: object) => Promise<object>
  resetSynonyms: () => Promise<object>
  getStopWords: () => Promise<string[]>
  updateStopWords: (stopWords: string[]) => Promise<EnqueuedUpdate>
  resetStopWords: () => Promise<EnqueuedUpdate>
  getRankingRules: () => Promise<string[]>
  updateRankingRules: (rankingRules: string[]) => Promise<EnqueuedUpdate>
  resetRankingRules: () => Promise<EnqueuedUpdate>
  getDistinctAttribute: () => Promise<string | null>
  updateDistinctAttribute: (
    distinctAttribute: string
  ) => Promise<EnqueuedUpdate>
  resetDistinctAttribute: () => Promise<EnqueuedUpdate>
  getAttributesForFaceting: () => Promise<string[]>
  updateAttributesForFaceting: (
    attributesForFaceting: string[]
  ) => Promise<EnqueuedUpdate>
  resetAttributesForFaceting: () => Promise<EnqueuedUpdate>
  getSearchableAttributes: () => Promise<string[]>
  updateSearchableAttributes: (
    searchableAttributes: string[]
  ) => Promise<EnqueuedUpdate>
  resetSearchableAttributes: () => Promise<EnqueuedUpdate>
  getDisplayedAttributes: () => Promise<string[]>
  updateDisplayedAttributes: (
    displayedAttributes: string[]
  ) => Promise<EnqueuedUpdate>
  resetDisplayedAttributes: () => Promise<EnqueuedUpdate>
  getAcceptNewFields: () => Promise<boolean>
  updateAcceptNewFields: (acceptNewFields: boolean) => Promise<EnqueuedUpdate>
}

export interface MeiliAxiosWrapperInterface {
  instance: AxiosInstance
  cancelTokenSource: CancelTokenSource
  get: <T = any, R = AxiosResponse<T>>(
    url: string,
    config?: AxiosRequestConfig
  ) => Promise<R>
  post: (<T = any>(
    url: string,
    data: IndexRequest,
    config?: AxiosRequestConfig
  ) => Promise<Index<T>>) &
    (<T = any, R = AxiosResponse<EnqueuedUpdate>>(
      url: string,
      data?: T,
      config?: AxiosRequestConfig
    ) => Promise<R>)
  put: <T = any, R = AxiosResponse<T>>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ) => Promise<R>
  patch: <T = any, R = AxiosResponse<T>>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ) => Promise<R>
  delete: <T = any, R = AxiosResponse<T>>(
    url: string,
    config?: AxiosRequestConfig
  ) => Promise<R>
}

/*
 ** ERROR HANDLER
 */

export interface MeiliSearchApiErrorInterface extends Error {
  name: string
  message: string
  stack?: string
  errorCode?: string
  errorType?: string
  errorLink?: string
  response?: MeiliSearchApiErrorResponse
  request?: MeiliSearchApiErrorRequest
}
export interface MeiliSearchApiErrorResponse {
  status?: number
  statusText?: string
  path?: string
  method?: string
  body?: object
}
export interface MeiliSearchApiErrorRequest {
  url?: string
  path?: string
  method?: string
}

export type MeiliSearchApiErrorConstructor = new (
  error: AxiosError,
  cachedStack?: string
) => void

export default MeiliSearch
