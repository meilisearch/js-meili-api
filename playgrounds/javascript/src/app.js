import { MeiliSearch } from '../../../src'

const config = {
  host: 'http://127.0.0.1:7700',
  apiKey: 'masterKey',
}

const client = new MeiliSearch(config)
const indexUid = 'movies'

const addDataset = async () => {
  await client.deleteIndex(indexUid)
  const task = await client.createIndex(indexUid)
  await client.index(indexUid).waitForTask(task.taskUid)

  const documents = await client.index(indexUid).getDocuments()

  const dataset = [
    { id: 1, title: 'Carol', genres: ['Romance', 'Drama'] },
    { id: 2, title: 'Wonder Woman', genres: ['Action', 'Adventure'] },
    { id: 3, title: 'Life of Pi', genres: ['Adventure', 'Drama'] },
    {
      id: 4,
      title: 'Mad Max: Fury Road',
      genres: ['Adventure', 'Science Fiction'],
    },
    { id: 5, title: 'Moana', genres: ['Fantasy', 'Action'] },
    { id: 6, title: 'Philadelphia', genres: ['Drama'] },
  ]
  if (documents.length === 0) {
    const task = await client.index(indexUid).addDocuments(dataset)
    await client.index(indexUid).waitForTask(task.taskUid)
  }
}

;(async () => {
  try {
    await addDataset()
    const indexes = await client.getRawIndexes()
    document.querySelector('.indexes').innerText = JSON.stringify(
      indexes,
      null,
      1
    )
    const resp = await client.index(indexUid).search('', {
      attributesToHighlight: ['title'],
    })
    console.log({ resp })
    console.log({ hit: resp.hits[0] })
    document.querySelector('.hits').innerText = JSON.stringify(
      resp.hits.map((hit) => hit.title),
      null,
      1
    )
    document.querySelector('.errors_title').style.display = 'none'
  } catch (e) {
    console.error(e)
    document.querySelector('.errors').innerText = JSON.stringify(e, null, 1)
  }
})()
