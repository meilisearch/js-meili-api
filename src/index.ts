/*
 * Bundle: Meilisearch / Indexes
 * Project: Meilisearch - Javascript API
 * Author: Quentin de Quelen <quentin@meilisearch.com>
 * Copyright: 2019, Meilisearch
 */

'use strict'

import { MeiliSearchTimeOutError } from './errors/meilisearch-timeout-error'
import MeiliAxiosWrapper from './meili-axios-wrapper'
import * as Types from './types'
import { sleep } from './utils'

class Index extends MeiliAxiosWrapper implements Types.Index {
  uid: string
  constructor(config: Types.Config, uid: string) {
    super(config)
    this.uid = uid
  }

  ///
  /// UPDATES
  ///

  /**
   * Get the informations about an update status
   * @memberof Index
   * @method getUpdateStatus
   */
  getUpdateStatus(updateId: number): Promise<Types.Update> {
    const url = `/indexes/${this.uid}/updates/${updateId}`

    return this.get(url)
  }

  async waitForPendingUpdate(
    updateId: number,
    {
      timeOutMs = 5000,
      intervalMs = 50,
    }: { timeOutMs?: number; intervalMs?: number } = {}
  ) {
    const startingTime = Date.now()
    while (Date.now() - startingTime < timeOutMs) {
      const response = await this.getUpdateStatus(updateId)
      if (response.status !== 'enqueued') return response
      await sleep(intervalMs)
    }
    throw new MeiliSearchTimeOutError(
      `timeout of ${timeOutMs}ms has exceeded on process ${updateId} when waiting for pending update to resolve.`
    )
  }

  /**
   * Get the list of all updates
   * @memberof Index
   * @method getAllUpdateStatus
   */
  getAllUpdateStatus(): Promise<Types.Update[]> {
    const url = `/indexes/${this.uid}/updates`

    return this.get(url)
  }

  ///
  /// SEARCH
  ///

  /**
   * Search for documents into an index
   * @memberof Index
   * @method search
   */
  search(
    query: string,
    options?: Types.SearchParams
  ): Promise<Types.SearchResponse> {
    const url = `/indexes/${this.uid}/search`

    const params: Types.SearchRequest = {
      q: query,
    }
    if (options) {
      if (options.offset) {
        params.offset = options.offset
      }
      if (options.limit) {
        params.limit = options.limit
      }
      if (options.attributesToRetrieve) {
        if (Array.isArray(options.attributesToRetrieve)) {
          params.attributesToRetrieve = options.attributesToRetrieve.join(',')
        } else {
          params.attributesToRetrieve = options.attributesToRetrieve
        }
      }

      if (options.attributesToCrop) {
        if (Array.isArray(options.attributesToCrop)) {
          params.attributesToCrop = options.attributesToCrop.join(',')
        } else {
          params.attributesToCrop = options.attributesToCrop
        }
      }
      if (options.cropLength) {
        params.cropLength = options.cropLength
      }
      if (options.attributesToHighlight) {
        if (Array.isArray(options.attributesToHighlight)) {
          params.attributesToHighlight = options.attributesToHighlight.join(',')
        } else {
          params.attributesToHighlight = options.attributesToHighlight
        }
      }
      if (options.filters) {
        params.filters = options.filters
      }

      if (options.matches) {
        params.matches = options.matches
      }
    }

    return this.get(url, {
      params,
      cancelToken: this.cancelTokenSource.token,
    })
  }

  ///
  /// INDEX
  ///
  /**
   * Show index information.
   * @memberof Index
   * @method show
   */
  show(): Promise<Types.IndexResponse> {
    const url = `/indexes/${this.uid}`

    return this.get(url)
  }

  /**
   * Update an index.
   * @memberof Index
   * @method updateIndex
   */
  updateIndex(data: Types.UpdateIndexRequest): Promise<Types.IndexResponse> {
    const url = `/indexes/${this.uid}`

    return this.put(url, data)
  }

  /**
   * Delete an index.
   * @memberof Index
   * @method deleteIndex
   */

  deleteIndex(): Promise<string> {
    const url = `/indexes/${this.uid}`

    return this.delete(url)
  }

  ///
  /// STATS
  ///

  /**
   * get stats of an index
   * @memberof Index
   * @method getStats
   */
  getStats(): Promise<Types.IndexStats> {
    const url = `/indexes/${this.uid}/stats`

    return this.get(url)
  }
  ///
  /// DOCUMENTS
  ///

  /**
   * get documents of an index
   * @memberof Index
   * @method getDocuments
   */
  getDocuments(options?: Types.GetDocumentsParams): Promise<Types.Document[]> {
    const url = `/indexes/${this.uid}/documents`
    let attr
    if (options && Array.isArray(options.attributesToRetrieve)) {
      attr = options.attributesToRetrieve.join(',')
    }

    return this.get(url, {
      params: {
        ...options,
        ...(attr ? { attributesToRetrieve: attr } : {}),
      },
    })
  }

  /**
   * Get one document
   * @memberof Index
   * @method getDocument
   */
  getDocument(documentId: string | number): Promise<Types.Document> {
    const url = `/indexes/${this.uid}/documents/${documentId}`

    return this.get(url)
  }

  /**
   * Add or replace multiples documents to an index
   * @memberof Index
   * @method addDocuments
   */
  addDocuments(
    documents: Types.Document[],
    options?: Types.AddDocumentParams
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents`

    return this.post(url, documents, {
      params: options,
    })
  }

  /**
   * Add or update multiples documents to an index
   * @memberof Index
   * @method updateDocuments
   */
  updateDocuments(
    documents: Types.Document[],
    options?: Types.AddDocumentParams
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents`

    return this.put(url, documents, {
      params: options,
    })
  }

  /**
   * Delete one document
   * @memberof Index
   * @method deleteDocument
   */
  deleteDocument(documentId: string | number): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents/${documentId}`

    return this.delete(url)
  }

  /**
   * Delete multiples documents of an index
   * @memberof Index
   * @method deleteDocuments
   */
  deleteDocuments(
    documentsIds: string[] | number[]
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents/delete-batch`

    return this.post(url, documentsIds)
  }

  /**
   * Delete all documents of an index
   * @memberof Index
   * @method deleteAllDocuments
   */
  deleteAllDocuments(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/documents`

    return this.delete(url)
  }

  ///
  /// SETTINGS
  ///

  /**
   * Retrieve all settings
   * @memberof Index
   * @method getSettings
   */
  getSettings(): Promise<Types.Settings> {
    const url = `/indexes/${this.uid}/settings`

    return this.get(url)
  }

  /**
   * Update all settings
   * Any parameters not provided will be left unchanged.
   * @memberof Index
   * @method updateSettings
   */
  updateSettings(settings: Types.Settings): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings`

    return this.post(url, settings)
  }

  /**
   * Reset settings.
   * @memberof Index
   * @method resetSettings
   */
  resetSettings(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings`

    return this.delete(url)
  }

  ///
  /// SYNONYMS
  ///

  /**
   * Get the list of all synonyms
   * @memberof Index
   * @method getSynonyms
   */
  getSynonyms(): Promise<object> {
    const url = `/indexes/${this.uid}/settings/synonyms`

    return this.get(url)
  }

  /**
   * Update the list of synonyms. Overwrite the old list.
   * @memberof Index
   * @method updateSynonyms
   */
  updateSynonyms(synonyms: object): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/synonyms`

    return this.post(url, synonyms)
  }

  /**
   * Reset the synonym list to be empty again
   * @memberof Index
   * @method resetSynonyms
   */
  resetSynonyms(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/synonyms`

    return this.delete(url)
  }

  ///
  /// STOP WORDS
  ///

  /**
   * Get the list of all stop-words
   * @memberof Index
   * @method getStopWords
   */
  getStopWords(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/stop-words`

    return this.get(url)
  }

  /**
   * Update the list of stop-words. Overwrite the old list.
   * @memberof Index
   * @method updateStopWords
   */
  updateStopWords(stopWords: string[]): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/stop-words`

    return this.post(url, stopWords)
  }

  /**
   * Reset the stop-words list to be empty again
   * @memberof Index
   * @method resetStopWords
   */
  resetStopWords(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/stop-words`

    return this.delete(url)
  }

  ///
  /// RANKING RULES
  ///

  /**
   * Get the list of all ranking-rules
   * @memberof Index
   * @method getRankingRules
   */
  getRankingRules(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/ranking-rules`

    return this.get(url)
  }

  /**
   * Update the list of ranking-rules. Overwrite the old list.
   * @memberof Index
   * @method updateRankingRules
   */
  updateRankingRules(rankingRules: string[]): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/ranking-rules`

    return this.post(url, rankingRules)
  }

  /**
   * Reset the ranking rules list to its default value
   * @memberof Index
   * @method resetRankingRules
   */
  resetRankingRules(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/ranking-rules`

    return this.delete(url)
  }

  ///
  /// DISTINCT ATTRIBUTE
  ///

  /**
   * Get the distinct-attribute
   * @memberof Index
   * @method getDistinctAttribute
   */
  getDistinctAttribute(): Promise<string | void> {
    const url = `/indexes/${this.uid}/settings/distinct-attribute`

    return this.get(url)
  }

  /**
   * Update the distinct-attribute.
   * @memberof Index
   * @method updateDistinctAttribute
   */
  updateDistinctAttribute(
    distinctAttribute: string
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/distinct-attribute`

    return this.post(url, distinctAttribute)
  }

  /**
   * Reset the distinct-attribute.
   * @memberof Index
   * @method resetDistinctAttribute
   */
  resetDistinctAttribute(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/distinct-attribute`

    return this.delete(url)
  }

  ///
  /// SEARCHABLE ATTRIBUTE
  ///

  /**
   * Get the searchable-attributes
   * @memberof Index
   * @method getSearchableAttributes
   */
  getSearchableAttributes(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/searchable-attributes`

    return this.get(url)
  }

  /**
   * Update the searchable-attributes.
   * @memberof Index
   * @method updateSearchableAttributes
   */
  updateSearchableAttributes(
    searchableAttributes: string[]
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/searchable-attributes`

    return this.post(url, searchableAttributes)
  }

  /**
   * Reset the searchable-attributes.
   * @memberof Index
   * @method resetSearchableAttributes
   */
  resetSearchableAttributes(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/searchable-attributes`

    return this.delete(url)
  }

  ///
  /// DISPLAYED ATTRIBUTE
  ///

  /**
   * Get the displayed-attributes
   * @memberof Index
   * @method getDisplayedAttributes
   */
  getDisplayedAttributes(): Promise<string[]> {
    const url = `/indexes/${this.uid}/settings/displayed-attributes`

    return this.get(url)
  }

  /**
   * Update the displayed-attributes.
   * @memberof Index
   * @method updateDisplayedAttributes
   */
  updateDisplayedAttributes(
    displayedAttributes: string[]
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/displayed-attributes`

    return this.post(url, displayedAttributes)
  }

  /**
   * Reset the displayed-attributes.
   * @memberof Index
   * @method resetDisplayedAttributes
   */
  resetDisplayedAttributes(): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/displayed-attributes`

    return this.delete(url)
  }

  ///
  /// ACCEPT NEW FIELDS
  ///

  /**
   * Get the accept-new-fields value.
   * @memberof Index
   * @method getAcceptNewFields
   */
  getAcceptNewFields(): Promise<boolean> {
    const url = `/indexes/${this.uid}/settings/accept-new-fields`

    return this.get(url)
  }

  /**
   * Update the accept-new-fields value.
   * @memberof Index
   * @method updateAcceptNewFields
   */
  updateAcceptNewFields(
    acceptNewFields: boolean
  ): Promise<Types.EnqueuedUpdate> {
    const url = `/indexes/${this.uid}/settings/accept-new-fields`

    return this.post(url, acceptNewFields)
  }
}

export { Index }
