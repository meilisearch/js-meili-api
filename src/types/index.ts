export * from "./settings.js";
export * from "./types.js";
