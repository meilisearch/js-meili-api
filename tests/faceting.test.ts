import { ErrorStatusCode } from '../src/types';
import {
  clearAllIndexes,
  config,
  BAD_HOST,
  MeiliSearch,
  getClient,
  dataset,
} from './utils/meilisearch-test-utils';

const index = {
  uid: 'movies_test',
};

jest.setTimeout(100 * 1000);

afterAll(() => {
  return clearAllIndexes(config);
});

describe.each([{ permission: 'Master' }, { permission: 'Admin' }])(
  'Test on faceting',
  ({ permission }) => {
    beforeEach(async () => {
      await clearAllIndexes(config);
      const client = await getClient('Master');
      const { taskUid } = await client.createIndex(index.uid);
      await client.waitForTask(taskUid);

      const { taskUid: docTask } = await client
        .index(index.uid)
        .addDocuments(dataset);
      await client.waitForTask(docTask);
    });

    test(`${permission} key: Get default faceting object`, async () => {
      const client = await getClient(permission);

      const response = await client.index(index.uid).getFaceting();

      expect(response).toMatchSnapshot();
    });

    test(`${permission} key: Update faceting settings`, async () => {
      const client = await getClient(permission);
      const newFaceting = {
        maxValuesPerFacet: 12,
        sortFacetValuesBy: { test: 'count' as 'count' },
      };
      const task = await client.index(index.uid).updateFaceting(newFaceting);
      await client.index(index.uid).waitForTask(task.taskUid);

      const response = await client.index(index.uid).getFaceting();

      expect(response).toMatchSnapshot();
    });

    test(`${permission} key: Update faceting at null`, async () => {
      const client = await getClient(permission);
      const task = await client
        .index(index.uid)
        .updateFaceting({ maxValuesPerFacet: null });
      await client.index(index.uid).waitForTask(task.taskUid);

      const response = await client.index(index.uid).getFaceting();

      expect(response).toMatchSnapshot();
    });

    test(`${permission} key: Reset faceting`, async () => {
      const client = await getClient(permission);
      await client
        .index(index.uid)
        .waitForTask(
          (
            await client
              .index(index.uid)
              .updateFaceting({ maxValuesPerFacet: 12 })
          ).taskUid,
        );
      const task = await client.index(index.uid).resetFaceting();
      await client.index(index.uid).waitForTask(task.taskUid);

      const response = await client.index(index.uid).getFaceting();

      expect(response).toMatchSnapshot();
    });
  },
);

describe.each([{ permission: 'Search' }])(
  'Test on faceting',
  ({ permission }) => {
    beforeEach(async () => {
      const client = await getClient('Master');
      const { taskUid } = await client.createIndex(index.uid);
      await client.waitForTask(taskUid);
    });

    test(`${permission} key: try to get faceting and be denied`, async () => {
      const client = await getClient(permission);
      await expect(
        client.index(index.uid).getFaceting(),
      ).rejects.toHaveProperty('cause.code', ErrorStatusCode.INVALID_API_KEY);
    });

    test(`${permission} key: try to update faceting and be denied`, async () => {
      const client = await getClient(permission);
      await expect(
        client.index(index.uid).updateFaceting({ maxValuesPerFacet: 13 }),
      ).rejects.toHaveProperty('cause.code', ErrorStatusCode.INVALID_API_KEY);
    });

    test(`${permission} key: try to reset faceting and be denied`, async () => {
      const client = await getClient(permission);
      await expect(
        client.index(index.uid).resetFaceting(),
      ).rejects.toHaveProperty('cause.code', ErrorStatusCode.INVALID_API_KEY);
    });
  },
);

describe.each([{ permission: 'No' }])('Test on faceting', ({ permission }) => {
  beforeAll(async () => {
    const client = await getClient('Master');
    const { taskUid } = await client.createIndex(index.uid);
    await client.waitForTask(taskUid);
  });

  test(`${permission} key: try to get faceting and be denied`, async () => {
    const client = await getClient(permission);
    await expect(client.index(index.uid).getFaceting()).rejects.toHaveProperty(
      'cause.code',
      ErrorStatusCode.MISSING_AUTHORIZATION_HEADER,
    );
  });

  test(`${permission} key: try to update faceting and be denied`, async () => {
    const client = await getClient(permission);
    await expect(
      client.index(index.uid).updateFaceting({ maxValuesPerFacet: 13 }),
    ).rejects.toHaveProperty(
      'cause.code',
      ErrorStatusCode.MISSING_AUTHORIZATION_HEADER,
    );
  });

  test(`${permission} key: try to reset faceting and be denied`, async () => {
    const client = await getClient(permission);
    await expect(
      client.index(index.uid).resetFaceting(),
    ).rejects.toHaveProperty(
      'cause.code',
      ErrorStatusCode.MISSING_AUTHORIZATION_HEADER,
    );
  });
});

describe.each([
  { host: BAD_HOST, trailing: false },
  { host: `${BAD_HOST}/api`, trailing: false },
  { host: `${BAD_HOST}/trailing/`, trailing: true },
])('Tests on url construction', ({ host, trailing }) => {
  test(`Test getFaceting route`, async () => {
    const route = `indexes/${index.uid}/settings/faceting`;
    const client = new MeiliSearch({ host });
    const strippedHost = trailing ? host.slice(0, -1) : host;
    await expect(client.index(index.uid).getFaceting()).rejects.toHaveProperty(
      'message',
      `Request to ${strippedHost}/${route} has failed`,
    );
  });

  test(`Test updateFaceting route`, async () => {
    const route = `indexes/${index.uid}/settings/faceting`;
    const client = new MeiliSearch({ host });
    const strippedHost = trailing ? host.slice(0, -1) : host;
    await expect(
      client.index(index.uid).updateFaceting({ maxValuesPerFacet: null }),
    ).rejects.toHaveProperty(
      'message',
      `Request to ${strippedHost}/${route} has failed`,
    );
  });

  test(`Test resetFaceting route`, async () => {
    const route = `indexes/${index.uid}/settings/faceting`;
    const client = new MeiliSearch({ host });
    const strippedHost = trailing ? host.slice(0, -1) : host;
    await expect(
      client.index(index.uid).resetFaceting(),
    ).rejects.toHaveProperty(
      'message',
      `Request to ${strippedHost}/${route} has failed`,
    );
  });
});
