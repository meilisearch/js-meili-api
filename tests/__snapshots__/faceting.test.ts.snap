// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on faceting Admin key: Get default faceting object 1`] = `
{
  "maxValuesPerFacet": 100,
  "sortFacetValuesBy": {
    "*": "alpha",
  },
}
`;

exports[`Test on faceting Admin key: Reset faceting 1`] = `
{
  "maxValuesPerFacet": 100,
  "sortFacetValuesBy": {
    "*": "alpha",
  },
}
`;

exports[`Test on faceting Admin key: Update faceting at null 1`] = `
{
  "maxValuesPerFacet": 100,
  "sortFacetValuesBy": {
    "*": "alpha",
  },
}
`;

exports[`Test on faceting Admin key: Update faceting settings 1`] = `
{
  "maxValuesPerFacet": 12,
  "sortFacetValuesBy": {
    "*": "alpha",
    "test": "count",
  },
}
`;

exports[`Test on faceting Master key: Get default faceting object 1`] = `
{
  "maxValuesPerFacet": 100,
  "sortFacetValuesBy": {
    "*": "alpha",
  },
}
`;

exports[`Test on faceting Master key: Reset faceting 1`] = `
{
  "maxValuesPerFacet": 100,
  "sortFacetValuesBy": {
    "*": "alpha",
  },
}
`;

exports[`Test on faceting Master key: Update faceting at null 1`] = `
{
  "maxValuesPerFacet": 100,
  "sortFacetValuesBy": {
    "*": "alpha",
  },
}
`;

exports[`Test on faceting Master key: Update faceting settings 1`] = `
{
  "maxValuesPerFacet": 12,
  "sortFacetValuesBy": {
    "*": "alpha",
    "test": "count",
  },
}
`;
